dkg/pedersen/handler.go

```diff
@@ -185,144 +195,112 @@ func (h *Handler) Stream(out mino.Sender, in mino.Receiver) error {
-	from mino.Address, out mino.Sender) error {
-
-	dela.Logger.Info().Msgf("%v is starting a DKG", h.me)
+func (h *Handler) start(start types.Start, deals, resps *list.List, from mino.Address,
+	out mino.Sender) error {
 
-	participants := start.GetAddresses()
-
-	if len(participants) != len(start.GetPublicKeys()) {
-		return xerrors.Errorf(
-			"there should be as many participants as "+
-				"pubKey: %d != %d", len(start.GetAddresses()),
-			len(start.GetPublicKeys()),
-		)
+	if len(start.GetAddresses()) != len(start.GetPublicKeys()) {
+		return xerrors.Errorf("there should be as many participants as "+
+			"pubKey: %d != %d", len(start.GetAddresses()), len(start.GetPublicKeys()))
 	}
-	n := len(participants)
-	ctx, cancel := context.WithTimeout(ctx, time.Duration(n*n)*time.Second)
-	defer cancel()
 
-	// 1. Create the DKG
-	d, err := pedersen.NewDistKeyGenerator(suite, h.privKey, start.GetPublicKeys(), start.GetThreshold())
+	// create the DKG
+	t := threshold.ByzantineThreshold(len(start.GetPublicKeys()))
+	d, err := pedersen.NewDistKeyGenerator(suite, h.privKey, start.GetPublicKeys(), t)
 	if err != nil {
 		return xerrors.Errorf("failed to create new DKG: %v", err)
 	}
+
 	h.dkg = d
 
-	// 2. Send my Deals to the other nodes
-	err = h.sendDeals(ctx, out, participants)
-	if err != nil {
-		return xerrors.Errorf("failed to send deals: %v", err)
-	}
+	h.startRes.SetParticipants(start.GetAddresses())
 
-	// 3. Process the incoming deals
-	err = h.receiveDeals(ctx, participants, from, out)
+	// asynchronously start the procedure. This allows for receiving messages
+	// in the main for loop in the meantime.
+	go h.doDKG(deals, resps, out, from)
+
+	return nil
+}
+
+// doDKG calls the subsequent DKG steps
+func (h *Handler) doDKG(deals, resps *list.List, out mino.Sender, from mino.Address) {
+	h.log.Info().Str("action", "deal").Msg(newState)
+	h.deal(out)
+
+	h.log.Info().Str("action", "respond").Msg(newState)
+	h.respond(deals, out)
+
+	h.log.Info().Str("action", "certify").Msg(newState)
+	err := h.certify(resps, out)
 	if err != nil {
-		return xerrors.Errorf("failed to receive deals: %v", err)
+		dela.Logger.Error().Msgf("failed to certify: %v", err)
+		return
 	}
 
-	h.startRes.SetParticipants(participants)
+	h.log.Info().Str("action", "finalize").Msg(newState)
 
-	// 4. Certify
-	err = h.certify(ctx)
+	// Send back the public DKG key
+	distKey, err := h.dkg.DistKeyShare()
 	if err != nil {
-		return xerrors.Errorf("failed to certify: %v", err)
+		dela.Logger.Error().Msgf("failed to get distr key: %v", err)
+		return
 	}
 
-	// 5. Announce the DKG public key
-	err = h.announceDkgPublicKey(out, from)
+	// Update the state before sending to acknowledgement to the
+	// orchestrator, so that it can process decrypt requests right away.
+	h.startRes.SetDistKey(distKey.Public())
+
+	h.Lock()
+	h.privShare = distKey.PriShare()
+	h.running = false
+	h.Unlock()
+
+	done := types.NewStartDone(distKey.Public())
+	err = <-out.Send(done, from)
```

Comment:
it seemed sort of natural (and easy) to put all of this in `announceDkgPublicKey` , given the different levels of abstraction.
What do you think ?

Reply:
@nkcr :innocent: